File: frgfp/_core/src/_ansatz.cpp
#include "_ansatz.hpp"
#include <vector>

namespace {
    struct Caches1D {
        std::vector<Eigen::MatrixXd> v0;
        std::vector<Eigen::MatrixXd> v1;
        std::vector<Eigen::MatrixXd> v2;
    };

    Caches1D build_poly_caches(int inv_dim, const Eigen::VectorXi& order, const Eigen::VectorXd& center, const Eigen::MatrixXd& grid) {
        int n_points = grid.rows();
        Caches1D c;
        c.v0.resize(inv_dim); c.v1.resize(inv_dim); c.v2.resize(inv_dim);
        
        for (int d = 0; d < inv_dim; ++d) {
            int ord = order(d);
            c.v0[d] = Eigen::MatrixXd::Zero(n_points, ord + 1);
            c.v1[d] = Eigen::MatrixXd::Zero(n_points, ord + 1);
            c.v2[d] = Eigen::MatrixXd::Zero(n_points, ord + 1);
            
            for (int p = 0; p < n_points; ++p) {
                double x_c = grid(p, d) - center(d);
                c.v0[d](p, 0) = 1.0;
                
                if (ord >= 1) {
                    c.v0[d](p, 1) = x_c;
                    c.v1[d](p, 1) = 1.0;
                }
                if (ord >= 2) {
                    c.v0[d](p, 2) = x_c * x_c;
                    c.v1[d](p, 2) = 2.0 * x_c;
                    c.v2[d](p, 2) = 2.0;
                }
                for (int k = 3; k <= ord; ++k) {
                    c.v0[d](p, k) = c.v0[d](p, k - 1) * x_c;
                    c.v1[d](p, k) = k * c.v0[d](p, k - 1);
                    c.v2[d](p, k) = k * (k - 1) * c.v0[d](p, k - 2);
                }
            }
        }
        return c;
    }

    Caches1D build_cheb_caches(int inv_dim, const Eigen::VectorXi& order, const Eigen::MatrixXd& interval, const Eigen::MatrixXd& grid) {
        int n_points = grid.rows();
        Caches1D c;
        c.v0.resize(inv_dim); c.v1.resize(inv_dim); c.v2.resize(inv_dim);
        
        for (int d = 0; d < inv_dim; ++d) {
            int ord = order(d);
            c.v0[d] = Eigen::MatrixXd::Zero(n_points, ord + 1);
            c.v1[d] = Eigen::MatrixXd::Zero(n_points, ord + 1);
            c.v2[d] = Eigen::MatrixXd::Zero(n_points, ord + 1);
            
            double min_val = interval(d, 0);
            double max_val = interval(d, 1);
            double scale = 2.0 / (max_val - min_val);

            for (int p = 0; p < n_points; ++p) {
                double y = scale * (grid(p, d) - min_val) - 1.0;

                c.v0[d](p, 0) = 1.0;
                if (ord >= 1) {
                    c.v0[d](p, 1) = y;
                    c.v1[d](p, 1) = scale;
                }
                for (int k = 2; k <= ord; ++k) {
                    c.v0[d](p, k) = 2.0 * y * c.v0[d](p, k - 1) - c.v0[d](p, k - 2);
                    c.v1[d](p, k) = 2.0 * scale * c.v0[d](p, k - 1) + 2.0 * y * c.v1[d](p, k - 1) - c.v1[d](p, k - 2);
                    c.v2[d](p, k) = 4.0 * scale * c.v1[d](p, k - 1) + 2.0 * y * c.v2[d](p, k - 1) - c.v2[d](p, k - 2);
                }
            }
        }
        return c;
    }

    // Flat coefficient index -> per-dimension multi-index, built once per call.
    std::vector<int> build_multi_index_table(int inv_dim, const Eigen::VectorXi& order, int num_coeffs) {
        std::vector<int> table(static_cast<size_t>(num_coeffs) * inv_dim, 0);
        for (int i = 0; i < num_coeffs; ++i) {
            int current_flat = i;
            for (int d = inv_dim - 1; d >= 0; --d) {
                const int base = order(d) + 1;
                table[static_cast<size_t>(i) * inv_dim + d] = current_flat % base;
                current_flat /= base;
            }
        }
        return table;
    }

    // Tensor-product evaluations shared by both ansaetze (cache-layout identical).

    Eigen::MatrixXd eval_basis_from_caches(const Caches1D& caches, int inv_dim,
                                           int n_points, int num_coeffs,
                                           const std::vector<int>& idx) {
        Eigen::MatrixXd M(n_points, num_coeffs);
        for (int i = 0; i < num_coeffs; ++i) {
            const int* mi = &idx[static_cast<size_t>(i) * inv_dim];
            for (int p = 0; p < n_points; ++p) {
                double val = 1.0;
                for (int d = 0; d < inv_dim; ++d) {
                    val *= caches.v0[d](p, mi[d]);
                }
                M(p, i) = val;
            }
        }
        return M;
    }

    std::vector<Eigen::MatrixXd> eval_grad_from_caches(const Caches1D& caches, int inv_dim,
                                                       int n_points, int num_coeffs,
                                                       const std::vector<int>& idx) {
        std::vector<Eigen::MatrixXd> grads(inv_dim, Eigen::MatrixXd::Zero(n_points, num_coeffs));
        for (int i = 0; i < num_coeffs; ++i) {
            const int* mi = &idx[static_cast<size_t>(i) * inv_dim];
            for (int p = 0; p < n_points; ++p) {
                for (int target_d = 0; target_d < inv_dim; ++target_d) {
                    double val = 1.0;
                    for (int d = 0; d < inv_dim; ++d) {
                        if (d == target_d) val *= caches.v1[d](p, mi[d]);
                        else val *= caches.v0[d](p, mi[d]);
                    }
                    grads[target_d](p, i) = val;
                }
            }
        }
        return grads;
    }

    std::vector<Eigen::MatrixXd> eval_hess_from_caches(const Caches1D& caches, int inv_dim,
                                                       int n_points, int num_coeffs,
                                                       const std::vector<int>& idx) {
        const int hess_dim = inv_dim * inv_dim;
        std::vector<Eigen::MatrixXd> hesss(hess_dim, Eigen::MatrixXd::Zero(n_points, num_coeffs));
        for (int i = 0; i < num_coeffs; ++i) {
            const int* mi = &idx[static_cast<size_t>(i) * inv_dim];
            for (int p = 0; p < n_points; ++p) {
                for (int d1 = 0; d1 < inv_dim; ++d1) {
                    for (int d2 = 0; d2 < inv_dim; ++d2) {
                        double val = 1.0;
                        for (int d = 0; d < inv_dim; ++d) {
                            if (d == d1 && d == d2) val *= caches.v2[d](p, mi[d]);
                            else if (d == d1 || d == d2) val *= caches.v1[d](p, mi[d]);
                            else val *= caches.v0[d](p, mi[d]);
                        }
                        hesss[d1 * inv_dim + d2](p, i) = val;
                    }
                }
            }
        }
        return hesss;
    }
}

namespace frgfp {
namespace core {

// --- Ansatz (default fused evaluation) ---

void Ansatz::evaluate_all(const Eigen::MatrixXd& grid,
                          Eigen::MatrixXd& M,
                          std::vector<Eigen::MatrixXd>& grads,
                          std::vector<Eigen::MatrixXd>& hesss) const {
    M = evaluate_basis(grid);
    grads = evaluate_basis_grad(grid);
    hesss = evaluate_basis_hess(grid);
}

// --- ChebyshevAnsatz ---

ChebyshevAnsatz::ChebyshevAnsatz(int inv_dim, const Eigen::VectorXi& order, const Eigen::MatrixXd& interval)
    : Ansatz(inv_dim, 0), m_order(order), m_interval(interval) {
    int total = 1;
    for (int i = 0; i < m_order.size(); ++i) total *= (m_order(i) + 1);
    this->m_num_coeffs = total;
}

Eigen::MatrixXd ChebyshevAnsatz::evaluate_basis(const Eigen::MatrixXd& grid) const {
    const Caches1D caches = build_cheb_caches(m_inv_dim, m_order, m_interval, grid);
    return eval_basis_from_caches(caches, m_inv_dim, grid.rows(), m_num_coeffs,
                                  build_multi_index_table(m_inv_dim, m_order, m_num_coeffs));
}

std::vector<Eigen::MatrixXd> ChebyshevAnsatz::evaluate_basis_grad(const Eigen::MatrixXd& grid) const {
    const Caches1D caches = build_cheb_caches(m_inv_dim, m_order, m_interval, grid);
    return eval_grad_from_caches(caches, m_inv_dim, grid.rows(), m_num_coeffs,
                                 build_multi_index_table(m_inv_dim, m_order, m_num_coeffs));
}

std::vector<Eigen::MatrixXd> ChebyshevAnsatz::evaluate_basis_hess(const Eigen::MatrixXd& grid) const {
    const Caches1D caches = build_cheb_caches(m_inv_dim, m_order, m_interval, grid);
    return eval_hess_from_caches(caches, m_inv_dim, grid.rows(), m_num_coeffs,
                                 build_multi_index_table(m_inv_dim, m_order, m_num_coeffs));
}

void ChebyshevAnsatz::evaluate_all(const Eigen::MatrixXd& grid,
                                   Eigen::MatrixXd& M,
                                   std::vector<Eigen::MatrixXd>& grads,
                                   std::vector<Eigen::MatrixXd>& hesss) const {
    // One cache build and one multi-index table feed all three evaluations.
    const Caches1D caches = build_cheb_caches(m_inv_dim, m_order, m_interval, grid);
    const int n_points = grid.rows();
    const std::vector<int> idx = build_multi_index_table(m_inv_dim, m_order, m_num_coeffs);
    M = eval_basis_from_caches(caches, m_inv_dim, n_points, m_num_coeffs, idx);
    grads = eval_grad_from_caches(caches, m_inv_dim, n_points, m_num_coeffs, idx);
    hesss = eval_hess_from_caches(caches, m_inv_dim, n_points, m_num_coeffs, idx);
}

// --- PolyAnsatz ---

PolyAnsatz::PolyAnsatz(int inv_dim, const Eigen::VectorXi& order, const Eigen::VectorXd& center)
    : Ansatz(inv_dim, 0), m_order(order), m_center(center) {
    int total = 1;
    for (int i = 0; i < m_order.size(); ++i) total *= (m_order(i) + 1);
    this->m_num_coeffs = total;
}

Eigen::MatrixXd PolyAnsatz::evaluate_basis(const Eigen::MatrixXd& grid) const {
    const Caches1D caches = build_poly_caches(m_inv_dim, m_order, m_center, grid);
    return eval_basis_from_caches(caches, m_inv_dim, grid.rows(), m_num_coeffs,
                                  build_multi_index_table(m_inv_dim, m_order, m_num_coeffs));
}

std::vector<Eigen::MatrixXd> PolyAnsatz::evaluate_basis_grad(const Eigen::MatrixXd& grid) const {
    const Caches1D caches = build_poly_caches(m_inv_dim, m_order, m_center, grid);
    return eval_grad_from_caches(caches, m_inv_dim, grid.rows(), m_num_coeffs,
                                 build_multi_index_table(m_inv_dim, m_order, m_num_coeffs));
}

std::vector<Eigen::MatrixXd> PolyAnsatz::evaluate_basis_hess(const Eigen::MatrixXd& grid) const {
    const Caches1D caches = build_poly_caches(m_inv_dim, m_order, m_center, grid);
    return eval_hess_from_caches(caches, m_inv_dim, grid.rows(), m_num_coeffs,
                                 build_multi_index_table(m_inv_dim, m_order, m_num_coeffs));
}

void PolyAnsatz::evaluate_all(const Eigen::MatrixXd& grid,
                              Eigen::MatrixXd& M,
                              std::vector<Eigen::MatrixXd>& grads,
                              std::vector<Eigen::MatrixXd>& hesss) const {
    // One cache build and one multi-index table feed all three evaluations.
    const Caches1D caches = build_poly_caches(m_inv_dim, m_order, m_center, grid);
    const int n_points = grid.rows();
    const std::vector<int> idx = build_multi_index_table(m_inv_dim, m_order, m_num_coeffs);
    M = eval_basis_from_caches(caches, m_inv_dim, n_points, m_num_coeffs, idx);
    grads = eval_grad_from_caches(caches, m_inv_dim, n_points, m_num_coeffs, idx);
    hesss = eval_hess_from_caches(caches, m_inv_dim, n_points, m_num_coeffs, idx);
}

} // namespace core
} // namespace frgfp

File: frgfp/_core/_ansatz.py
"""
Polynomial and Chebyshev ansatz definitions for the frgfp core module.

This module exposes the user-facing basis-expansion classes. Their inputs are
validated and normalized in Python, while the actual basis, gradient, and
Hessian evaluations are delegated to the optimised ``_core_cpp`` Eigen backend.
"""

import numpy as np
from typing import Union, Tuple, List
from . import _core_cpp

def _validate_common_ansatz_inputs(inv_dim: int, order: Union[int, Tuple[int, ...]]):
    """
    Validates inv_dim and order, returning a normalized order tuple.
    """
    if not isinstance(inv_dim, int) or isinstance(inv_dim, bool):
        raise TypeError(f"The 'inv_dim' parameter must be an integer. Received: {type(inv_dim).__name__}")
    if inv_dim < 1:
        raise ValueError(f"The 'inv_dim' must be a strictly positive integer. Received: {inv_dim}")

    if inv_dim == 1 and isinstance(order, (int, np.integer)):
        order = (order,)

    if not isinstance(order, tuple):
        raise TypeError(f"The 'order' parameter must be an int (for 1D) or a tuple of ints. Received: {type(order).__name__}")
    if len(order) != inv_dim:
        raise ValueError(f"The length of the 'order' tuple ({len(order)}) must match 'inv_dim' ({inv_dim}).")
    for i, ord_val in enumerate(order):
        if not isinstance(ord_val, (int, np.integer)) or isinstance(ord_val, bool):
            raise TypeError(f"All elements in 'order' must be integers. Found type {type(ord_val).__name__} at index {i}.")
        if ord_val < 0:
            raise ValueError(f"Polynomial orders cannot be negative. Found {ord_val} at index {i}.")

    return order

class ChebyshevAnsatz(_core_cpp.ChebyshevAnsatz):
    """
    Multidimensional Chebyshev polynomial ansatz defining the basis expansion.

    Constructs a highly optimized, orthogonal Chebyshev basis relying on an 
    Eigen C++ backend. The implementation maps the physical intervals to `[-1, 1]` 
    and evaluates the basis utilizing a stable recursive relation internally.

    Parameters
    ----------
    inv_dim : int
        The number of invariants (dimensions).
    order : int or tuple of int
        The maximum polynomial order for each invariant.
    interval : tuple of float or tuple of tuples of float
        The physical interval for expansion, which will be linearly mapped to the 
        standard `[-1, 1]` domain internally.
    """
    def __init__(self, inv_dim: int, order: Union[int, Tuple[int, ...]], interval: Union[Tuple[float, float], Tuple[Tuple[float, float], ...]]):
        """
        Initialize a Chebyshev basis expansion.

        Parameters
        ----------
        inv_dim : int
            The number of invariants (dimensions).
        order : int or tuple of int
            The maximum polynomial order for each invariant.
        interval : tuple of float or tuple of tuples of float
            The physical expansion interval per dimension, mapped internally to
            the standard ``[-1, 1]`` domain.
        """
        order = _validate_common_ansatz_inputs(inv_dim, order)

        if inv_dim == 1 and isinstance(interval, tuple) and len(interval) == 2 and isinstance(interval[0], (int, float, np.number)):
            interval = (interval,)

        if not isinstance(interval, tuple):
            raise TypeError(f"The 'interval' parameter must be a tuple. Received: {type(interval).__name__}")
        if len(interval) != inv_dim:
            raise ValueError(f"The length of the 'interval' tuple ({len(interval)}) must match 'inv_dim' ({inv_dim}).")
        
        for i, bounds in enumerate(interval):
            if not isinstance(bounds, tuple) or len(bounds) != 2:
                raise ValueError(f"Each dimension in 'interval' must be a tuple of length 2 (min, max). Error at dimension {i}: {bounds}")
            if not all(isinstance(x, (int, float, np.number)) and not isinstance(x, bool) for x in bounds):
                raise TypeError(f"Interval bounds must be numeric (int or float). Error at dimension {i}: {bounds}")
            if bounds[0] >= bounds[1]:
                raise ValueError(f"In 'interval', minimum values must be strictly less than maximum values. Error at dimension {i}: {bounds[0]} >= {bounds[1]}")

        order_arr = np.array(order, dtype=np.int32)
        interval_arr = np.array(interval, dtype=np.float64)

        super().__init__(inv_dim, order_arr, interval_arr)


    @property
    def inv_dim(self) -> int:
        """The dimensionality of the invariant space."""
        return super().inv_dim

    @property
    def num_coeffs(self) -> int:
        """The total number of coefficients in the tensor-product basis."""
        return super().num_coeffs

    def evaluate_basis(self, grid: np.ndarray) -> np.ndarray:
        """
        Evaluates the multidimensional basis functions on the given grid.

        Parameters
        ----------
        grid : np.ndarray
            A 2D array of shape `(N_points, inv_dim)` containing the grid coordinates.

        Returns
        -------
        np.ndarray
            A 2D array of shape `(N_points, N_coeffs)` containing the basis function values.
        """
        return super().evaluate_basis(grid)

    def evaluate_basis_grad(self, grid: np.ndarray) -> List[np.ndarray]:
        """
        Evaluates the gradient of the basis functions on the given grid.

        Parameters
        ----------
        grid : np.ndarray
            A 2D array of shape `(N_points, inv_dim)` containing the grid coordinates.

        Returns
        -------
        list of np.ndarray
            A list of length `inv_dim`, where each element is a 2D array of shape 
            `(N_points, N_coeffs)` representing the basis gradient components.
        """
        return super().evaluate_basis_grad(grid)

    def evaluate_basis_hess(self, grid: np.ndarray) -> List[np.ndarray]:
        """
        Evaluates the Hessian of the basis functions on the given grid.

        Parameters
        ----------
        grid : np.ndarray
            A 2D array of shape `(N_points, inv_dim)` containing the grid coordinates.

        Returns
        -------
        list of np.ndarray
            A list of length `inv_dim ** 2` (row-major order), where each element is 
            a 2D array of shape `(N_points, N_coeffs)` representing the basis Hessian components.
        """
        return super().evaluate_basis_hess(grid)


class PolyAnsatz(_core_cpp.PolyAnsatz):
    """
    Standard multidimensional polynomial ansatz defining a local Taylor-like basis.

    Constructs a polynomial expansion evaluated around a designated central 
    point (the expansion minimum). The basis functions follow the form 
    $(x - x_0)^n$. Evaluations are strictly processed in the Eigen C++ backend.

    Parameters
    ----------
    inv_dim : int
        The number of invariants (dimensions).
    order : int or tuple of int
        The maximum polynomial order for each invariant.
    center : float or tuple of float
        The local expansion point $x_0$ for each invariant. 
    """
    def __init__(self, inv_dim: int, order: Union[int, Tuple[int, ...]], center: Union[float, Tuple[float, ...]]):
        """
        Initialize a polynomial (Taylor) basis expansion.

        Parameters
        ----------
        inv_dim : int
            The number of invariants (dimensions).
        order : int or tuple of int
            The maximum polynomial order for each invariant.
        center : float or tuple of float
            The local expansion point ``x_0`` for each invariant.
        """
        order = _validate_common_ansatz_inputs(inv_dim, order)

        if inv_dim == 1 and isinstance(center, (int, float, np.number)) and not isinstance(center, bool):
            center = (center,)

        if not isinstance(center, tuple):
            raise TypeError(f"The 'center' parameter must be a float (for 1D) or a tuple of floats. Received: {type(center).__name__}")
        if len(center) != inv_dim:
            raise ValueError(f"The length of the 'center' tuple ({len(center)}) must match 'inv_dim' ({inv_dim}).")
        for i, c_val in enumerate(center):
            if not isinstance(c_val, (int, float, np.number)) or isinstance(c_val, bool):
                raise TypeError(f"All elements in 'center' must be numeric. Found type {type(c_val).__name__} at index {i}.")

        order_arr = np.array(order, dtype=np.int32)
        center_arr = np.array(center, dtype=np.float64)

        super().__init__(inv_dim, order_arr, center_arr)


    @property
    def inv_dim(self) -> int:
        """The dimensionality of the invariant space."""
        return super().inv_dim

    @property
    def num_coeffs(self) -> int:
        """The total number of coefficients in the tensor-product basis."""
        return super().num_coeffs


    def evaluate_basis(self, grid: np.ndarray) -> np.ndarray:
        """
        Evaluates the multidimensional basis functions on the given grid.

        Parameters
        ----------
        grid : np.ndarray
            A 2D array of shape `(N_points, inv_dim)` containing the grid coordinates.

        Returns
        -------
        np.ndarray
            A 2D array of shape `(N_points, N_coeffs)` containing the basis function values.
        """
        return super().evaluate_basis(grid)

    def evaluate_basis_grad(self, grid: np.ndarray) -> List[np.ndarray]:
        """
        Evaluates the gradient of the basis functions on the given grid.

        Parameters
        ----------
        grid : np.ndarray
            A 2D array of shape `(N_points, inv_dim)` containing the grid coordinates.

        Returns
        -------
        list of np.ndarray
            A list of length `inv_dim`, where each element is a 2D array of shape 
            `(N_points, N_coeffs)` representing the basis gradient components.
        """
        return super().evaluate_basis_grad(grid)

    def evaluate_basis_hess(self, grid: np.ndarray) -> List[np.ndarray]:
        """
        Evaluates the Hessian of the basis functions on the given grid.

        Parameters
        ----------
        grid : np.ndarray
            A 2D array of shape `(N_points, inv_dim)` containing the grid coordinates.

        Returns
        -------
        list of np.ndarray
            A list of length `inv_dim ** 2` (row-major order), where each element is 
            a 2D array of shape `(N_points, N_coeffs)` representing the basis Hessian components.
        """
        return super().evaluate_basis_hess(grid)